File: pythonLeetcode/206.反转链表.py
#
# @lc app=leetcode.cn id=206 lang=python3
#
# [206] 反转链表
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# 解包
#   def reverseList(self, head: ListNode) -> ListNode:
#     cur, prev = head, None
#     while cur:
#       cur.next, prev, cur = prev, cur, cur.next
#     return prev

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        fakeNode = None
        while head:
            tmp = head.next
            head.next = fakeNode
            fakeNode = head
            head = tmp
        return fakeNode
        
# @lc code=end

File: StudyLeetcodeProject/StudyLeetcodeProject/Finished/链表/翻转链表/206.反转链表.swift
//
//  206.反转链表.swift
//  StudyLeetcodeProject
//
//  Created by beliefduan on 2022/8/16.
//

import Foundation


class Solution206 {
    func reverseList(_ head: ListNode?) -> ListNode? {
        if head == nil || head?.next == nil {
            return head
        }
        let l = reverseList(head?.next)
        head?.next?.next = head
        head?.next = nil
        return l
    }
    func reverseList1(_ head: ListNode?) -> ListNode? {
        var head = head
        var cur:ListNode?
        while head != nil {
            let tmp = head?.next
            head?.next = cur
            cur = head
            head = tmp
        }
        return cur
        
    }
}
